cmd/prom-migrator/main.go

```diff
@@ -0,0 +1,194 @@
+)
+
+type config struct {
+	name               string
+	mint               int64
+	maxt               int64
+	maxBlockSizeBytes  int64
+	maxBlockSize       string
+	readURL            string
+	writeURL           string
+	progressMetricName string
+	progressMetricURL  string
+	progressEnabled    bool
+	readerAuth         utils.Auth
+	writerAuth         utils.Auth
+}
+
+func main() {
+	conf := new(config)
+	parseFlags(conf, os.Args[1:])
+	if err := log.Init(log.Config{Format: "logfmt", Level: "debug"}); err != nil {
+		_, _ = fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+	if err := validateConf(conf); err != nil {
+		log.Error("msg", "could not parse flags", "error", err)
+		os.Exit(1)
+	}
+	log.Info("msg", fmt.Sprintf("%v+", conf))
+	if err := utils.SetAuthStore(utils.Read, conf.readerAuth.ToHTTPClientConfig()); err != nil {
+		log.Error("msg", "could not set read-auth in authStore", "error", err)
+		os.Exit(1)
+	}
+	if err := utils.SetAuthStore(utils.Write, conf.readerAuth.ToHTTPClientConfig()); err != nil {
+		log.Error("msg", "could not set write-auth in authStore", "error", err)
+		os.Exit(1)
+	}
+
+	planConfig := &plan.Config{
+		Mint:                conf.mint,
+		Maxt:                conf.maxt,
+		BlockSizeLimitBytes: conf.maxBlockSizeBytes,
+	}
+	planner, proceed, err := plan.Init(planConfig)
+	if err != nil {
+		log.Error("msg", "could not create plan", "error", err)
+		os.Exit(2)
+	}
+	if !proceed {
+		os.Exit(0)
+	}
+
+	var (
+		readErrChan  = make(chan error)
+		writeErrChan = make(chan error)
+		sigBlockRead = make(chan *plan.Block)
+	)
+	cont, cancelFunc := context.WithCancel(context.Background())
+	read, err := reader.New(cont, conf.readURL, planner, sigBlockRead)
+	if err != nil {
+		log.Error("msg", "could not create reader", "error", err)
+		os.Exit(2)
+	}
+	write, err := writer.New(cont, conf.writeURL, conf.progressMetricName, conf.name, sigBlockRead)
+	if err != nil {
+		log.Error("msg", "could not create writer", "error", err)
+		os.Exit(2)
+	}
+
+	read.Run(readErrChan)
+	write.Run(writeErrChan)
+
+loop:
+	for {
+		select {
+		case err = <-readErrChan:
+			if err != nil {
+				cancelFunc()
+				log.Error("msg", fmt.Errorf("running reader: %w", err).Error())
+				os.Exit(2)
+			}
+		case err, ok := <-writeErrChan:
+			cancelFunc() // As in any ideal case, the reader will always exit normally first.
+			if ok {
+				log.Error("msg", fmt.Errorf("running writer: %w", err).Error())
+				os.Exit(2)
+			}
+			break loop
+		}
+	}
+
+	log.Info("msg", "migration successfully carried out")
+	log.Info("msg", "exiting!")
+}
+
+func parseFlags(conf *config, args []string) {
+	flag.StringVar(&conf.name, "migration-name", migrationJobName, "Name for the current migration that is to be carried out. "+
+		"It corresponds to the value of the label 'job' set inside the progress-metric-name.")
+	flag.Int64Var(&conf.mint, "mint", 0, "Minimum timestamp (in seconds) for carrying out data migration. (inclusive)")
```

Comment:
I would prefer these flags in Seconds be different conf variables than the mint fields and the convert code fill in the mint fields. Overwritting a var like that is a recipe for bugs down the line
```suggestion
	flag.Int64Var(&conf.mintSec, "mint", 0, "Minimum timestamp (in seconds) for carrying out data migration. (inclusive)")
```